public/gameInfo.php

```diff
@@ -307,7 +307,7 @@ class='btn break-words whitespace-normal leading-normal'
             && !$isBeatenHardcore
             && !$isBeatenSoftcore
         ) {
-            $beatenGameRetVal = testBeatenGame($gameID, $user, true);
+            $beatenGameRetVal = testBeatenGame($gameID, $user);
```

Comment:
One thing I am curious about (sorry if it is answered elsewhere in the PR).

Other than visiting a game page directly, how will the self-healing of beaten games work?

Example:
Game X has 100 achievements. 10 of them are Progression.
50 players have earned _only_ those 10 Progression achievements. Today, they have the beaten game award.
Tomorrow, a developer sets an 11th achievement as Progression.

Currently in V4, as `UpdatePlayerPoints` slowly crawls through the userbase, it will eventually correct everyone's beaten game awards over time, even if they don't visit the game page again or do a manual rescore.

Is there a similar automated mechanism in this PR to capture these cases?

Reply:
Technically this is addressed ad-hoc by listening on AchievementPublished events (achievement set version changes) which will mark all affected player_games to be updated.

During player_games metrics updates the badges will be re-evaluated for eligibility, too, which will act accordingly and remove the badge.

How the update of player games based on achievement set version changes will work in detail is still open - most likely it will be a scheduled task that takes care of scheduling all the player games marked for update (update_status) or triggered by a user logging in.

The previous "loops" that were crawling games and players do not scale well, thus I removed them completely. The StaticData table's references to those will be dropped, too, eventually.

Reply:
At the very least, we should be able to easily detect if the award is outdated when visiting a game page. Take a look at this screenshot:

![Screenshot 2023-09-24 at 11 46 09 AM](https://github.com/RetroAchievements/RAWeb/assets/3984985/db846261-657a-4b3b-becb-1c8e6cfc402c)

In the screenshot above, the user got the beaten game award, then a developer marked a new achievement as Progression. When this occurs, is it reasonable to fire off an event to update the player's game metrics before the page renders? Otherwise, we end up with situations where clearly I haven't met the beaten criteria, but the page still says I've beaten the game.

Reply:
Ideally this should've been caught fairly soon after the new achievement was published or its type changed so it will cascade down into all the players' games metrics to then remove any ineligibly awards.
I'd like to not cause any writes-on-visit side-effects anymore - this stuff has to happen in the back in a timely manner.

What i noticed though is that the type does have an event but no listener yet.
Instead of a listener I think it'd would make sense to make the achievement type part of the achievement version hash 🤔

Reply:
I agree that side effects on page visit are undesirable.

No strong opinion on the best way to handle this- it does make sense that on a type change some kind of event will cause a cascading update so player's don't really notice this (in some kind of realistic timeframe). We can also update the UI for the progress component to not display it to the player (we have all the info we need on the page to pull off this kind of optimistic visual update).

If we have a strategy for handling the data updates in the back-end, I can take care of the aforementioned UI piece. How do you think we should move forward?

Reply:
I like that - if we know on the game level that there's no beaten award to be had then there's no need to even check for a player's progress.

Reply:
If you can take care of the back-end, I will open a separate PR this weekend to adjust the UI in that component so the front-end is taken care of.

Reply:
Perfect, thanks! I'm already working on adding the achievement type to the achievement set version hash.

Reply:
Addressed in https://github.com/RetroAchievements/RAWeb/pull/1874/commits/239907663b08f50598a8e8d622455560d052ea15

Reply:
UI piece should be addressed in https://github.com/RetroAchievements/RAWeb/pull/1900.

Reply:
Fantastic! With the aggregated metrics in place this may become even simpler by just reading from the player's game data directly.

Reply:
I've merged https://github.com/RetroAchievements/RAWeb/pull/1900, which removes the potential `testBeatenGame()` call on page visit when aggregate queries are enabled and updates the progress status label on an optimistic basis. I think it's safe to mark this feedback as resolved.